### PR TITLE
identities that have a base identity are not sorted Fixes #234

### LINKED ISSUE #234: identities that have a base identity are not sorted
In YANG when a set of individual identities have a common base identity, these individual identities are gathered under the base, but their order is non deterministic, as they are extracted from a Go map.

This has 2 negative consequences
* YGOT `generate` command can produce a different JSON schema each time it it run because the order of identities changes, even though there are no real changes
* the order of these identities will be different to the enum entries in the generated code (which are sorted alphabetically). Therefore is is not possible to reference enums in the generated code by the order of the identities in the JSON schema

### PR DESCRIPTION
**Pull request by SeanCondon**
See issue #234

### CODE REVIEW

**Comment by SeanCondon** [comment]
Hi @wenovus - do you mind taking a look at this one please?

File: pkg/yang/identity_test.go [hunk 1]
```diff
@@ -388,19 +388,19 @@ var treeTestCases = []identityTestCase{
 				module: "base",
 				name:   "GREATGRANDFATHER",
 				values: []string{
-					"GRANDFATHER",
-					"GREATUNCLE",
+					"BROTHER", // Order is by hierarchy first then alphabetically
```
Comment (wenovus) [inline pkg/yang/identity_test.go:391]: This rule doesn't seem intuitive to me. If we're going to order the names I think it should be by the name only.

**Review by wenovus — CHANGES_REQUESTED** [review]

File: pkg/yang/identity.go [hunk 2]
```diff
@@ -169,6 +170,9 @@ func (ms *Modules) resolveIdentities() []error {
 
 				// Build up a list of direct children of this identity.
 				base.Identity.Values = append(base.Identity.Values, i.Identity)
+				sort.Slice(base.Identity.Values, func(j, k int) bool {
+					return base.Identity.Values[j].Name < base.Identity.Values[k].Name
+				})
```
[the remark this answers is not in the document]
Comment (wenovus) [reply pkg/yang/identity.go:175]: I added a PR to clarify the comment in this code block: https://github.com/openconfig/goyang/pull/236

File: pkg/yang/identity_test.go (hunk 1 above)
Comment (SeanCondon) [reply pkg/yang/identity_test.go]: now that the sort is done on the 2nd stage the order is alphabetical only

File: pkg/yang/identity.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (SeanCondon) [reply pkg/yang/identity.go]: Thanks @wenovus - this is a better job - the sort happens only once per base identity

**Review by wenovus — APPROVED** [review]

**Comment by SeanCondon** [comment]
> Now this is _technically_ a backwards-incompatible change. Only the order is affected, but I'm wary of just introducing this change now that we're at version 1.0.0.
> 
> Sorry about the very late comment -- is it inconvenient for you to sort them downstream on your own instead of depending on goyang to sort it?

Hi @wenovus - I can't really see how it is backward in-compatible - there used to be 2 outputs of enums in `generated.go`:

1. the generated go code - the enums were always ordered alphabetically and now will not change
2. the JSON schema - the enums were never ordered, and would change order every time you ran it - so every time you ran `generate` it was backward in-compatible. Now with this patch running `generate` the enums are ordered alphabetically  it will be another variation of this order - just that it will be consistent with the previous runs order and with the generated code (1)

The only way I think it can be considered a problem is that someone was depending on the order to be different every time (while this was likely, there was no guarantee of this)

**Comment by wenovus** [comment]
Ah I see, this is actually a bug fix that makes the code deterministic. Sorry I should've read the issue you opened. Thanks for the clarification!

**Comment by SeanCondon** [comment]
@wenovus Thanks for merging. BTW - would it be possible to release this? Is that a manual process?

**Pull request merged**